### ISSUE TITLE
a11y: Improve message delivery status reporting

### ISSUE THREAD

**Issue by pvagner** [deltachat/deltachat-desktop#6605]
- Operating System: tested on Linux and Windows
- Delta Chat Version: 2.57
- Expected behavior: Improve outgoing message delivery status reporting for screen reader users
- Actual behavior: As explained in the notes below this part is now accessible but I'd like you to consider improving it even further
- Steps to reproduce the problem:
    * Launch deltachat desktop with a screen reader running
    * Navigate into a chat
    * Send a few messages for testing paying attention how the screen reader is reporting delivery status for each message
    * Use tab or shift+tab to navigate to focus the messages timeline
    * Make sure the screen reader is switched into focus mode
    * Use up and down arrow keys to navigate over the messages you have just sent and also please read my notes below

As I am using deltachat more and more I have decided to try improving its translation into slovak. I have used transifex, I have understood the same language strings are used on all the platforms including desktop, android and IOS.

After doing this I am thinking if there might be a way to improve message delivery status reporting for screen reader users even more. I can't say this part is not accessible but making it a bit more convenient would certainly be a nice bonus.

The first thing is that I find it too verbose when saying:

* Delivery status: Error
* Delivery status: Sending
* Delivery status: Draft
* Delivery status: Delivered
* Delivery status: Read
* Invalid delivery status

If you are okay with that I would recommend to drop the part before the colon so it becomes just "Sending", "Delivered", "Read" and "Draft". An exception might be "Delivery error" because plain error is too generic. I find it over verbosed when hearing this with each message. This point applies to both the desktop and android, but since the same strings are used for both I believe it's enough to bring it up once.

The second point is more detailed so here is my explanation of it:

When reading outgoing messages with a screen reader there is message delivery status such as sending, sent and similar.
The parent element has role="status" attribute set so it's reported with a screen reader whenever it changes.
This is mainly usefull for the most recent outgoing message and it's working well on both the platforms I am able to test windows and linux.

It's somewhat inconvenient for older outgoing messages because the content inside element with role="status" is not included for the accessible name calculation by rendering engine. Thus when using up and down arrow key to navigate in the list of messages as implemented according to the roving tabIndex pattern the message delivery status is not reported. Of course there are browse mode screen reader features that allow reading the message delivery status but I find it less convenient as I have to read the message content and its surrounding details line by line.

Would it be feasible to set role="status" for the last item only or would that cause unnecessary rerenders causing performance issues?
Or instead of using role="status" could the aria-live be used directly similar how it's already used for announcing selected message for replying?

**Comment by WofWca** [comment]
To summarize:
- "Delivery status: Delivered" and other such strings are too verbose
- Delivery status is not reported when focusing a message.

With the first thing I agree, I also brought it up some day. "Delivery status" should be the label of the `role="status"` element, and the contents should be just the "Delivered" or "Read". This will require changes on all DC clients.

Your suggestion for the second problem I am not so sure about. Rendering speed is not a big deal. But what I'm hesitant about is inconsistency: not really a likely scenario, but if you send two messages in quick succession, and the first one gets an error, you'd want that to get announced.
And overall: are the screen-readers actually doing the right thing by not announcing `role="status"` elements?

**Comment by pvagner** [comment]
> With the first thing I agree, I also brought it up some day. "Delivery status" should be the label of the `role="status"` element, and the contents should be just the "Delivered" or "Read". This will require changes on all DC clients.

I'm thinking that if there is nothing focusable inside the role="status" element then there is no need to label it as it will never be announced by screen readers except of when explicitly reading by line, by word and similar.



> Your suggestion for the second problem I am not so sure about. Rendering speed is not a big deal. But what I'm hesitant about is inconsistency: not really a likely scenario, but if you send two messages in quick succession, and the first one gets an error, you'd want that to get announced.

Oh I haven't accounted for this. Then perhaps doing it with live regions might work.
Or duplicating the visually hidden text for older messages might also fix reporting when navigating in the list although that looks cluttered.



> And overall: are the screen-readers actually doing the right thing by not announcing `role="status"` elements?

My gues is that this is electron not including it in the accessible name calculation for the focused element rather than screen readers ignoring it as I can reproduce it on both Windows with NVDA and Linux with Orca. Webkit-gtk on linux when testing the tauri build does it too.

**Comment by pvagner** [comment]
@WofWca 
Thanks for your patience with me.

For my second point where the delivery status is not reported for older messages by some trial and error I have tested that using `aria-live='polite'` directly instead of `role='status'` makes it work.
A diff is below.

However I'd need some hints how to deal with language strings if I would like to shorten them.

Should I hand edit file _locale/en.xml only or should I try to edit all the xml files inside the _locale folder? Should I make corresponding change for the android app too?


Here is my attempt at fixing this:

```diff
diff --git a/packages/frontend/src/components/message/MessageMetaData.tsx b/packages/frontend/src/components/message/MessageMetaData.tsx
index 2a57dfa94..8cf866cef 100644
--- a/packages/frontend/src/components/message/MessageMetaData.tsx
+++ b/packages/frontend/src/components/message/MessageMetaData.tsx
@@ -105,19 +105,19 @@ export default function MessageMetaData(props: Props) {
         !shouldHideDeliveryStatus(chatType, status)) ||
         error !== null) && (
         <div className='delivery-status-wrapper'>
-          {/* The main point of `role='status'` here is to let the user know
-          that their message has been sent or delievered
-          right after they they send it.
+          {/* The main point of `aria-live='polite'` here is to let the user
+          know that their message has been sent or delivered
+          right after they send it.
+          We used to have `role='status'` instead, but
+          that was causing issues reporting delivery status for older messages.
           We want at least some indication of something happening
           after they press "Enter".
           But this is also useful to announce when the message has been read.
 
           Note that this this applies to _all_ loaded messages
           and not just the last one. */}
-          {/* TODO a11y: we should probably apply `aria-label='Delivery status'`
-          and change the contents to `Delivered` (or `Error` or whatever). */}
           <div
-            role='status'
+            aria-live='polite' aria-atomic={true}
             className={classNames(
               'status-icon',
               error !== null ? 'error' : status
```

**Comment by WofWca** [comment]
> Should I hand edit file _locale/en.xml only or should I try to edit all the xml files inside the _locale folder? Should I make corresponding change for the android app too?

The source strings are located in the Delta Chat Android repo. That would be the place where this needs to change ideally. DC Desktop merely copy-pastes the translations from Transifex, with its `tx` CLI.

**Comment by pvagner** [comment]
> The source strings are located in the Delta Chat Android repo. That would be the place where this needs to change ideally.

These particular strings are referenced by their ID so I can imagine this might work.

However most of the calls to the tx function use the string it-self within deltachat desktop.

How is this supposed to work when adding new strings?

**Comment by pvagner** [comment]
> For my second point where the delivery status is not reported for older messages by some trial and error I have tested that using `aria-live='polite'` directly instead of `role='status'` makes it work.

Hmmm, I discovered that the issue with these live regions is that most of them are spoken on various other occasions when not really needed such as when opening the app into an existing conversation. Many delivery statuses are read along with edited and other things.

Anyway it's not related to this issue so perhaps using this would be an improvement anyway.

I'm thinking aloud that perhaps to improve these we might add a single visually hidden element with `role="status"` and use that to report important status messages as they happen. When reopening the app or scrolling a lot this unwanted reporting of multiple live regions won't happen.

**Comment by WofWca** [comment]
> However most of the calls to the tx function use the string it-self within deltachat desktop.

You mean that we pass the to-be-displayed string to `tx`, instead of a string ID? It's not the case, I'm not sure where you're getting that from.

**Comment by WofWca** [comment]
> most of them are spoken on various other occasions when not really needed such as when opening the app into an existing conversation

Umm I don't think that's supposed to happen. After such a live region gets inserted, its content doesn't change, so there is nothing to announce.
But indeed it does happen on Orca, e.g. with reactions...

**Comment by pvagner** [comment]
I will have to test this with NVDA on windows but on linux the steps definatelly showcase what I mean:

* Start deltachat
* Open a new conversation
* Write a few messages like three or four
* Edit all of them
* Close deltachat and open it again.
As it launches the second time the recent conversation is loaded and "Edited" is reported multiple times as the content populates.

It's why I'm thinking considering consolidating screen reader announcements to single aria live region might be usefull thing for the future.

**Comment by WofWca** [comment]
This also happens on Chromium. IDK why `role="status" would be different from `aria-live="polite"` in this regard...

```js
setInterval(() => {
  live = document.createElement('span')
  live.ariaLive = 'polite'
  live.textContent = 'test'
	wrap = document.createElement('div')
	wrap.appendChild(live)
  document.body.appendChild(wrap)
}, 3000)
```

**Comment by pvagner** [comment]
Oh I'm sorry, I think chromium does it too and it does not matter if it's a result of `role='status'` or `aria-live='polite'`. It's just that I have noticed it while testing yesterday as an addition to my original issue.

My suggestion would be to polish and apply this now and for the future consider reworking these screen reader announcements to the single live region if you can agree with that.

**Comment by WofWca** [comment]
> reworking these screen reader announcements to the single live region if you can agree with that.

TBH I'm not excited about that. Maybe I'm misunderstanding what you're actually proposing, but that sounds like we'd need quite some imperative code to achieve that. All this to work around screen-reader bugs (or rather our misunderstanding of how screen-readers are supposed to work)? IDK IDK

Where the path forward is clear to me is the "delivery status" string split.

**Comment by pvagner** [comment]
I have finally been able to test this with NVDA on windows and on windows both the browsers are working the same way similar to how firefox is working under linux. I can't figure it on my own so I have asked people posting to the orca email list to help clearing up my position on this.
Until nothing better is found I think we just need to use aria-live instead of `role='status'` and shorten the strings and hope for the fixes elsewhere.

I will look into the android app tommorow.

Thanks and greetings

Peter
